e2e/vaultsecrets/vaultsecrets_test.go

```diff
@@ -0,0 +1,196 @@
+	"time"
+
+	e2e "github.com/hashicorp/nomad/e2e/e2eutil"
+	"github.com/hashicorp/nomad/e2e/v3/jobs3"
+	"github.com/hashicorp/nomad/e2e/v3/namespaces3"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/shoenig/test"
+	"github.com/shoenig/test/must"
+	"github.com/shoenig/test/wait"
+)
+
+const ns = "vault-secrets"
+
+func TestVaultSecrets(t *testing.T) {
+	// use a random suffix to encapsulate test keys, polices, etc.
+	// for cleanup from vault
+	testID := uuid.Generate()[0:8]
+	secretsPath := "secrets-" + testID
+	pkiPath := "pki-" + testID
+	secretValue := uuid.Generate()
+	secretKey := secretsPath + "/data/myapp"
+	pkiCertIssue := pkiPath + "/issue/nomad"
+	policyID := "access-secrets-" + testID
+
+	t.Cleanup(namespaces3.Create(t, ns))
+
+	// configure KV secrets engine
+	// Note: the secret key is written to 'secret-###/myapp' but the kv2 API
+	// for Vault implicitly turns that into 'secret-###/data/myapp' so we
+	// need to use the longer path for everything other than kv put/get
+	e2e.MustCommand(t, "vault secrets enable -path=%s kv-v2", secretsPath)
+	e2e.CleanupCommand(t, "vault secrets disable %s", secretsPath)
+	e2e.MustCommand(t, "vault kv put %s/myapp key=%s", secretsPath, secretValue)
+	e2e.MustCommand(t, "vault secrets tune -max-lease-ttl=1m %s", secretsPath)
+
+	// configure PKI secrets engine
+	e2e.MustCommand(t, "vault secrets enable -path=%s pki", pkiPath)
+	e2e.CleanupCommand(t, "vault secrets disable %s", pkiPath)
+	e2e.MustCommand(t, "vault write %s/root/generate/internal "+
+		"common_name=service.consul ttl=1h", pkiPath)
+	e2e.MustCommand(t, "vault write %s/roles/nomad "+
+		"allowed_domains=service.consul "+
+		"allow_subdomains=true "+
+		"generate_lease=true "+
+		"max_ttl=1m", pkiPath)
+	e2e.MustCommand(t, "vault secrets tune -max-lease-ttl=1m %s", pkiPath)
+
+	// we can't set an empty policy in our job, so write a bogus policy that
+	// doesn't have access to any of the paths we're using
+	writePolicy(t, policyID, "./input/policy-bad.hcl", testID)
+
+	submission, cleanJob := jobs3.Submit(t,
+		"./input/secrets.nomad",
+		jobs3.Namespace(ns),
+		jobs3.Detach(),
+		jobs3.ReplaceInJobSpec("TESTID", testID),
+		jobs3.ReplaceInJobSpec("DEPLOYNUMBER", "FIRST"),
+	)
+	t.Cleanup(cleanJob)
+	jobID := submission.JobID()
+
+	// job doesn't have access to secrets, so they can't start
+	err := e2e.WaitForAllocStatusExpected(jobID, ns, []string{"pending"})
+	must.NoError(t, err, must.Sprint("expected pending allocation"))
+
+	// we should get a task event about why they can't start
+	expect := fmt.Sprintf("Missing: vault.read(%s), vault.write(%s", secretKey, pkiCertIssue)
+	allocID := submission.AllocID("group")
+
+	must.Wait(t, wait.InitialSuccess(
+		wait.ErrorFunc(func() error {
+			out, err := e2e.Command("nomad", "alloc", "status", "-namespace", ns, allocID)
+			if err != nil {
+				return err
+			}
+			if !strings.Contains(out, expect) {
+				return fmt.Errorf("test for alloc status failed; got:\n%v", out)
+			}
```

Comment:
Oh yeah, I saw those, but thought perhaps overkill for this case, since the expected string is so specific.

I went for it anyway, found a decently well-targeted thing `AllocTaskEventsForJob()` that uses those more granular helpers, and found a lil bug in it to fix too!

How's [c1afabf](https://github.com/hashicorp/nomad/pull/19152/commits/c1afabfbf369dfcf0d466024259828d5513df6df) look?

Reply:
LGTM! :shipit: